src/lib/gadgets/arithmetic.ts

```diff
@@ -0,0 +1,39 @@
+import { Field } from '../core.js';
+import { Provable } from '../provable.js';
+import { rangeCheck32 } from './range-check.js';
+
+export { divMod32, addMod32 };
+
+function divMod32(n: Field) {
+  if (n.isConstant()) {
+    let nBigInt = n.toBigInt();
+    let q = nBigInt / (1n << 32n);
+    let r = nBigInt - q * (1n << 32n);
+    return {
+      remainder: new Field(r),
+      quotient: new Field(q),
+    };
+  }
+
+  let qr = Provable.witness(Provable.Array(Field, 2), () => {
+    let nBigInt = n.toBigInt();
+    let q = nBigInt / (1n << 32n);
+    let r = nBigInt - q * (1n << 32n);
+    return [new Field(q), new Field(r)];
+  });
+  let [quotient, remainder] = qr;
```

Comment:
woah wtf

Reply:
fascinating, the inferred return type seems to be fully determined by the callback and result type, and completely overrides the `Provable<T>` type. So this is inferred to be `[Field, Field]` but then `Provable.Array(Field, 2)` doesn't match that.

This can probably be changed in either of two ways -- tweaking the type signature of `witness()`, or making `Provable.Array()` infer a tuple. The second would make it harder to use in some circumstances though.

We should probably separate "tuple" and "array", you can already do the former with `provableTuple` and this works:
```ts
let [quotient, remainder] = Provable.witness(
    provableTuple([Field, Field]),
    () => { ... });
```